tdm_ingestion/storage/tdmq.py

```diff
@@ -19,17 +20,17 @@ def create_from_json(cls, json: Dict):
         return CachedStorage(
             import_class(client['class']).create_from_json(client['args']))
 
-    def _idempotent_create(self, obj: Source):
-        if obj._id not in self._cache:
-            logging.debug('querying if source %s exists', obj._id)
-            if self.client.sources_count(query={'id': obj._id}) <= 0:
+    def _idempotent_create_source(self, obj: Source):
+        if obj.id_ not in self._cache:
+            logger.debug('querying if source %s exists', obj.id_)
+            if self.client.sources_count(query={'id': obj.id_}) <= 0:
```

Comment:
Looks like a race condition.  A concurrent process could create the source after your query.  I suggest you try to create it and catch the error if it already exists.

Reply:
I should have solved this in the new commits. Now it just creates the source without checking the existence before, and if it was already present, the server returns a duplicate error. I noticed another error that I should have solved: if the creation of the source fails, the time series would fail as well, so it should be better to skip the time series creation.